store/src/java/com/zimbra/cs/milter/MilterHandler.java

```diff
@@ -354,6 +355,33 @@ private void SMFIC_Rcpt() throws ServiceException {
             connection.send(new MilterPacket(SMFIR_TEMPFAIL));
             return;
         }
+        try { // ZBUG-2165: prevent forwarding to non-permitted distribution list
+            ZimbraLog.milter.trace("evaluating forwarding to non-permitted distribution list");
+            final Account account = prov.getAccountByName(rcpt);
+            if (account != null) {
+                ZimbraLog.milter.trace("account: %s", account.getName());
+                final String prefMailForwardingAddress = account.getPrefMailForwardingAddress();
+                ZimbraLog.milter.trace("prefMailForwardingAddress: %s", prefMailForwardingAddress);
+                if (prefMailForwardingAddress != null) {
+                    final String[] addresses = prefMailForwardingAddress.split(",");
+                    for (final String a : addresses) {
+                        final String address = a.trim();
+                        ZimbraLog.milter.trace("address: %s", address);
+                        if (!address.isEmpty() && prov.isDistributionList(address)) {
+                            final Group group = prov.getGroupBasic(Key.DistributionListBy.name, address);
+                            ZimbraLog.milter.trace("group: %s", group);
+                            if (group != null && !accessMgr.canDo(account, group, User.R_sendToDistList, false)) {
+                                ZimbraLog.milter.debug("Recipient %s is not allowed to forward to this distribution list: %s", account.getName(), group.getName());
+                                SMFIR_ReplyCode("571", "571 Recipient " + account.getName() + " is not allowed to forward to this distribution list: " + group.getName());
+                                return;
+                            }
+                        }
+                    }
+                }
+            }
+        } catch (Exception e) {
+            ZimbraLog.milter.warn("failed to evaluate forwarding to non-permitted distribution list");
```

Comment:
Please `catch` specific exception not the `Exception` class.

Reply:
Done.

I was going to catch the `ServiceException` class, which is the specific exception to be expected here, but then I realized that such class is already part of the `throw` clause on the parent `SMFIC_Rcpt()` method, so no point in catching it anywhere in this code. I completely removed the `try{}catch(){}` block instead.

Reply:
Note: this changes the behavior of this addition to fail, instead of continuing, if an exception is thrown while processing forwarding addresses.

Reply:
@wilfredo-alfaro You have get rid of `try` `catch` blocks but can you please address @desouzas concern.

Reply:
This is fine. It behaves the same as any ServiceException thrown in the section below.